tests/upgrades/scenario_workers.py

```diff
@@ -0,0 +1,55 @@
+import json
+import os
+
+import pytest
+from automation_tools.satellite6.hammer import set_hammer_config
+from fabric.api import env
+
+from robottelo.config import settings
+
+
+_json_file = 'upgrade_workers.json'
+
+
+def save_worker_id(test_name, worker_id):
+    data = {}
+    if os.path.exists(_json_file):
+        with open(_json_file) as rjson:
+            data = json.load(rjson)
+    with open(_json_file, 'w') as wjson:
+        # Removing the parameter name from test name before save
+        test_name = test_name.split('[')[0] if '[' in test_name else test_name
+        data.update({test_name: worker_id})
+        wjson.seek(0)
+        json.dump(data, wjson)
+
+
+def get_worker_id(test_name):
+    with open(_json_file) as rjson:
+        data = json.load(rjson)
+        return data.get(test_name)
+
+
+@pytest.fixture(autouse=True)
+def save_pretest_worker_id(request, worker_id):
+    """Saves the worker id of pre_upgrade test in json"""
+    if request.node.get_closest_marker('pre_upgrade'):
+        save_worker_id(request.node.name, worker_id)
+
+
+@pytest.fixture(autouse=True)
+def set_posttest_worker_id(request):
+    """Retrieves worker_id of pre_upgrade and sets the same to run post_upgrade"""
+    post_upgrade = request.node.get_closest_marker('post_upgrade')
+    if post_upgrade:
+        depend_test = post_upgrade.kwargs.get('depend_on')
+        if depend_test:
+            pre_test_name = depend_test.__name__
+            settings.configure()
+            worker_int = int(get_worker_id(pre_test_name).replace('gw', ''))
+            settings.server.hostname = settings.server.hostnames[worker_int]
```

Comment:
this direct assignment won't reliably work because the facade caches accessed values.

Reply:
I have tested this by creating a spcial test module and it works :)

Also, I am reinitiating the configuration on line number 48, so that refresh everything.

Reply:
I don't believe configure does anything anymore. The facade is mapping that call to a lambda that returns None.

https://github.com/SatelliteQE/robottelo/blob/41f3c4863f32c89195853f4bd07834523b47108c/robottelo/config/facade.py#L399

Reply:
Though it's caching the accessed values we are alternating the server name and it works fine!

Here is the Test Module I created:

```
from upgrade_tests import post_upgrade
from upgrade_tests import pre_upgrade
from nailgun import entities

org_id = 1


class Testscene1:

    prod_name = 'test11'

    @pre_upgrade
    def test_pre_1(self):
        prod = entities.Product(name=self.prod_name, organization=org_id).create()
        assert prod.name == self.prod_name

    @post_upgrade(depend_on=test_pre_1)
    def test_post_1(self):
        prod = entities.Product(organization=org_id).search(query={'search': f'name={self.prod_name}'})
        assert prod[0].name == self.prod_name


class Testscene2:

    prod_name = 'test22'

    @pre_upgrade
    def test_pre_2(self):
        prod = entities.Product(name=self.prod_name, organization=org_id).create()
        assert prod.name == self.prod_name

    @post_upgrade(depend_on=test_pre_2)
    def test_post_2(self):
        prod = entities.Product(organization=org_id).search(query={'search': f'name={self.prod_name}'})
        assert prod[0].name == self.prod_name


class Testscene3:

    prod_name = 'test33'

    @pre_upgrade
    def test_pre_3(self):
        prod = entities.Product(name=self.prod_name, organization=org_id).create()
        assert prod.name == self.prod_name

    @post_upgrade(depend_on=test_pre_3)
    def test_post_3(self):
        prod = entities.Product(organization=org_id).search(query={'search': f'name={self.prod_name}'})
        assert prod[0].name == self.prod_name


class Testscene4:

    prod_name = 'test44'

    @pre_upgrade
    def test_pre_4(self):
        prod = entities.Product(name=self.prod_name, organization=org_id).create()
        assert prod.name == self.prod_name

    @post_upgrade(depend_on=test_pre_4)
    def test_post_4(self):
        prod = entities.Product(organization=org_id).search(query={'search': f'name={self.prod_name}'})
        assert prod[0].name == self.prod_name

```

Reply:
@JacobCallahan Do you have any full-proof way to set the server name?

Reply:
To make a lasting change, I had to modify the value stored in the proxy's cache itself.
https://github.com/SatelliteQE/robottelo/blob/master/pytest_fixtures/xdist.py#L17-L19

Reply:
@JacobCallahan The suggested solution doesn't seem working for me, but the existing approach works well! I did negative testing as well and it shows the failures as expected.

Do you think I should implmemt the same allignment you have done ? Also, the facade is going away soon, I think we can merge this without wasting much effort on a thing that will move away. What do you think?

Reply:
I didn't try to set the server hostname to proxy btw, just tried to reset the cache and kept the existing way of setting the server hostname! Also, I am setting one sat for each worker so I don't need alignments that you have done in xdist.py.

@mshriver Thoughts ?

Reply:
I think you might be lucking out with timing in your tests, but I can't reliably say for sure that this will work when used in conjunction with everything else.

I won't hold on that, but please note the behavior in automation runs after this is merged.

@mshriver can still weigh in.

Reply:
@mshriver Ping !

Reply:
@mshriver @JacobCallahan Pings!!!

Reply:
Pong 🙃
We'll roll in Mike tomorrow

Reply:
@JacobCallahan Fixed this ! Now setting the server hostname in cache ! Also, tested this and its working great !